### PR TITLE
Better style tab

### PR DESCRIPTION
**Pull request by jrmi**

### CODE REVIEW

File: public/locales/en/translation.json [hunk 1]
```diff
@@ -835,5 +829,79 @@
         "file-placeholder": "Drop here Excel file to generate tree (Existing tree will be replaced)."
       }
     }
+  },
+  "style-editor":{
+    "polygon": "Polygone",
+    "line-type": "Ligne",
+    "point": "Point",
+    "other": "Autre type (${type})",
```
Comment (dtrucs) [inline public/locales/en/translation.json]: fr locales

File: src/components/react-admin/ColorPicker.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  position: 'fixed',
+  top: '0px',
+  right: '0px',
+  bottom: '0px',
+  left: '0px',
+};
+
+const presetColors = [];
+
+const ColorPicker = ({ value = '#ccccccff', onChange, disabled }) => {
```
Comment (dtrucs) [inline src/components/react-admin/ColorPicker.js:48]: a default value to `onChange` could be accurate to avoid app to crash if not filled.
Or change l.61 to `onChange?.(newColor.hex)`;

File: src/components/react-admin/ColorPicker.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      onChange(newColor.hex);
+    },
+    [onChange],
+  );
+
+  return (
+    <>
+      <div
+        className={classes.color}
+        onClick={() => !disabled && setShowPicker(!showPicker)}
```
Comment (dtrucs) [inline src/components/react-admin/ColorPicker.js:70]:
```suggestion
        onClick={() => !disabled && setShowPicker(prevShowPicked => !prevShowPicker)}
```

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/CategorizeValue.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+  React.useEffect(() => {
+    const getValueList = async () => {
+      if (prevField === field) return;
+      setTooManyValues(false);
+      const result = await getValuesOfProperty(field);
+
+      if (result.length > maxCategories) {
+        setTooManyValues(true);
+        return;
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/Style/CategorizeValue.js:40]: Maybe you have to check if this component is still mounted before executing a setState

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/CategorizeValue.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+              <Component
+                value={value}
+                onChange={onChange}
+              />
+            )}
+          </Field>
+          <span>
+            {category.name ?
+              category.name :
+              translate('style-editor.categorize.empty-category')}
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/Style/CategorizeValue.js:91]:
```suggestion
            {category.name || translate('style-editor.categorize.empty-category')}
```

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorListField.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  },
+});
+
+const ColorListField = ({ value, onChange }) => {
+  const classes = useStyles();
+
+  const handleColorChange = index => newValue => {
+    const newColorList = [...value];
+    newColorList[index] = newValue;
+    onChange(newColorList);
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorListField.js:30]: This and followers
```
    onChange?.(newColorList);
```

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorListField.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    newColorList.push(randomColor());
+    onChange(newColorList);
+  };
+
+
+  const removeColor = index => () => {
+    const newColorList = [...value];
+    newColorList.splice(index, 1);
+    onChange(newColorList);
+  };
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorListField.js:44]: Maybe you could wrap each handler below by an useCallbak

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorStyleField.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+import styles from './styles';
+
+const useStyles = makeStyles(styles);
+
+const ColorStyleField = ({ path, fields, getValuesOfProperty }) => {
+  const classes = useStyles();
+  const translate = useTranslate();
+
+  const [defaultSeed] = React.useState(Math.floor((Math.random() * 100000) + 1));
+  const [defaultValue] = React.useState([randomColor()]);
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorStyleField.js:27]: I'm not sure if this is relevant but you can put a function as argument of the useState to avoid computations every renders https://reactjs.org/docs/hooks-reference.html#lazy-initial-state

```suggestion
  const [defaultSeed] = React.useState(() => Math.floor((Math.random() * 100000) + 1));
  const [defaultValue] = React.useState(() => [randomColor()]);
```

File: src/components/react-admin/Condition.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+
+import { Field } from 'react-final-form';
+
+const Condition = ({ when, is, children }) => (
```
Comment (dtrucs) [inline src/components/react-admin/Condition.js:5]: Why not `if` instead of `when`?

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorStyleField.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+        {fields && (
+        <>
+          <SelectInput
+            source={`${path}.field`}
+            helperText="style-editor.field-help"
+            style={{ minWidth: '20em', margin: '1em 0' }}
+            label="style-editor.field"
+            choices={fields
+              .filter(field => [1, 2, 3].includes(field.data_type))
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorStyleField.js:60]: What are `1, 2, 3`? Maybe you should declare theses values by using capital-named const

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ValueListField.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  const classes = useStyles();
+
+  const handleValueChange = index => event => {
+    const newValueList = [...value];
+    const newValue = parseInt(event.target.value, 10);
+    if (Number.isNaN(newValue)) {
+      newValueList[index] = 0;
+    } else {
+      newValueList[index] = newValue;
+    }
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ValueListField.js:36]:
```suggestion
    newValueList[index] = Math.round(event.target.value) || 0;
```

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/styles.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
+const styles = {
+  configLine: {
+    '& header': {
+      borderRadius: '0.5em',
+      display: 'flex',
+      alignItems: 'center',
+      padding: '0 1em',
+      backgroundColor: '#eee',
+      // marginBottom: '1em',
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/Style/styles.js:9]: to remove

File: src/modules/RA/DataLayer/components/tabs/StyleTab/StyleEditor.js [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        </Condition>
+        <Condition when={`${path}.map_style_type`} is="circle">
+          <WizardCircle
+            path={path}
+            fields={fields}
+            getValuesOfProperty={getValuesOfProperty}
+          />
+        </Condition>
+        <Condition when={`${path}.map_style_type`} is="symbol">
+          <p>To be done</p>
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/StyleEditor.js:144]: Locale? or WIP?

**Review by dtrucs — CHANGES_REQUESTED** [review]

File: src/modules/RA/DataLayer/components/tabs/StyleTab/StyleEditor.js (hunk 13 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/StyleEditor.js:144]: Wip, will fix in another release.

File: src/components/react-admin/ColorPicker.js (hunk 3 above)
Comment (jrmi) [reply src/components/react-admin/ColorPicker.js:70]: Done

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/CategorizeValue.js (hunk 5 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/Style/CategorizeValue.js]: Done

File: src/components/react-admin/Condition.js (hunk 9 above)
Comment (jrmi) [reply src/components/react-admin/Condition.js:5]: I keep when as discussed.

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorStyleField.js (hunk 10 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorStyleField.js]: Done

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ValueListField.js (hunk 11 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ValueListField.js]: Done

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/styles.js (hunk 12 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/Style/styles.js]: done

File: src/components/react-admin/ColorPicker.js (hunk 2 above)
Comment (jrmi) [reply src/components/react-admin/ColorPicker.js]: Done

File: src/components/react-admin/ColorPicker.js (hunk 3 above)
Comment (jrmi) [reply src/components/react-admin/ColorPicker.js]: done

File: src/modules/RA/DataLayer/components/tabs/StyleTab/StyleEditor.js [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        />
+      </div>
+
+      {styleType === 'fill' && <h1>{translate('style-editor.geometry')} {translate('style-editor.polygon')}</h1>}
+      {styleType === 'line' && <h1>{translate('style-editor.geometry')} {translate('style-editor.line-type')}</h1>}
+      {styleType === 'circle' && <h1>{translate('style-editor.geometry')} {translate('style-editor.point')}</h1>}
+
+      {!['fill', 'line', 'circle'].includes(styleType) && (
+        <h2>{translate('style-editor.other', { type: styleType })}</h2>
+      )}
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/StyleEditor.js:49]: Is this relevant to change the tag when the type is unknown?
If not, maybe you can avoid some LOC
```jsx
<h1>
  {['fill', 'line', 'circle'].includes(styleType) 
    ? <>{translate('style-editor.geometry')}</>
    : <>{translate('style-editor.other', { type: styleType })}</>
  }
  {styleType === 'fill' && <>{translate('style-editor.polygon')}</>}
  {styleType === 'line' && <>{translate('style-editor.line')}</>}
  {styleType === 'circle' && <>{translate('style-editor.point')}</>}
</h1>
```

File: src/modules/RA/DataLayer/components/tabs/StyleTab/index.js [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    input: { value: fields },
+  } = useField('fields');
+
+  const { input: { value: mainStyle } } = useField('main_style');
+
+  const getValuesOfProperty = React.useCallback(property =>
+    dataProvider('PROPERTY_VALUES', RES_DATASOURCE, { id: sourceId, property }),
+  [dataProvider, sourceId]);
+
+  if (geomType === undefined || !sourceId) {
```
Comment (dtrucs) [inline src/modules/RA/DataLayer/components/tabs/StyleTab/index.js:38]: Now we can use the nullish coalescing operator for this kind of condition
`if (geomType ?? !sourceId) {`

**Review by dtrucs — CHANGES_REQUESTED** [review]

File: src/modules/RA/DataLayer/components/tabs/StyleTab/index.js (hunk 15 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/index.js]: Done

File: src/modules/RA/DataLayer/components/tabs/StyleTab/StyleEditor.js (hunk 14 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/StyleEditor.js]: Indeed. Done !

File: src/components/react-admin/ColorPicker.js [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+  position: 'absolute',
+  zIndex: '2',
+};
+
+const cover = {
+  position: 'fixed',
+  top: '0px',
+  right: '0px',
+  bottom: '0px',
+  left: '0px',
```
[the remark this answers is not in the document]
Comment (jrmi) [reply src/components/react-admin/ColorPicker.js]: Done

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorListField.js (hunk 6 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorListField.js:18]: Default value added instead. Done

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorListField.js (hunk 7 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorListField.js:31]: Not possible here as we have function that return function.

File: src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorStyleField.js (hunk 8 above)
Comment (jrmi) [reply src/modules/RA/DataLayer/components/tabs/StyleTab/Style/ColorStyleField.js]: Done

**Review by dtrucs — APPROVED** [review]

**Review by CFarcy — APPROVED** [review]
